### PR TITLE
🧟‍♂️  Dexpi Prepare allowing push of instances

### PR DESCRIPTION
**Pull request by nikokaoja**
## Overview
- Added bundle of tranformers for dexpi under `.prepare.dexpi()`
- Checks for multi typed instances added to query library
- Better handling of Undefined value types (removing namespace)
- extended default prefixes to account for dexpi, qudt, iodd

### Succesfull loading of DEXPI data model and instances
Screenshot 2024-12-08 at 10 57 16
Screenshot 2024-12-08 at 10 57 08

### CODE REVIEW

File: cognite/neat/_store/_base.py [hunk 1]
```diff
@@ -267,7 +277,7 @@ def _read_via_class_entity(
 
         # get property types to guide process of removing or not namespaces from results
         property_types = InformationAnalysis(self.rules).property_types(class_entity)
-
+        print(property_types)
```
Comment (nikokaoja) [inline cognite/neat/_store/_base.py:280]:
```suggestion
```

File: tests/tests_integration/test_session/test_graph_flow.py [hunk 2]
```diff
@@ -76,12 +77,56 @@ def test_classic_to_dms(self, cognite_client: CogniteClient, data_regression: Da
...
+
     @staticmethod
     def _standardize_instance(instance: InstanceApply) -> dict[str, Any]:
         if not isinstance(instance, InstanceApply):
             raise ValueError(f"Expected InstanceApply, got {type(instance)}")
         for source in instance.sources:
             for value in source.properties.values():
-                if isinstance(value, list):
+                if isinstance(value, list) and all(isinstance(v, dict) for v in value):
+                    value = sorted(value, key=lambda x: x["externalId"])
```
Comment (doctrino) [inline tests/tests_integration/test_session/test_graph_flow.py:129]: This has no effect as value is not used. In the below, value is sorted in place.

File: tests/tests_integration/test_session/test_graph_flow.py [hunk 3]
```diff
@@ -76,12 +77,56 @@ def test_classic_to_dms(self, cognite_client: CogniteClient, data_regression: Da
...
+        rules = neat._state.data_model.last_unverified_rule[1].rules
+        rules.metadata.created = "2024-09-19T00:00:00Z"
+        rules.metadata.updated = "2024-09-19T00:00:00Z"
+
+        neat.verify()
+
+        neat.convert("dms")
+        neat.set.data_model_id(("dexpi_playground", "DEXPI", "v1.3.1"))
+
+        if True:
```
Comment (doctrino) [inline tests/tests_integration/test_session/test_graph_flow.py:98]: Why this conditional? it serves no purpose.

File: tests/tests_integration/test_session/test_graph_flow.py [hunk 4]
```diff
@@ -76,12 +77,56 @@ def test_classic_to_dms(self, cognite_client: CogniteClient, data_regression: Da
...
+            instances = []
+
+        remove_linking_in_rules(neat._state.data_model.last_verified_dms_rules[1])
+        rules_str = neat.to.yaml(format="neat")
+
+        rules_dict = yaml.safe_load(rules_str)
+        data_regression.check(
+            {
+                "rules": rules_dict,
+                "instances": [],
```
Comment (doctrino) [inline tests/tests_integration/test_session/test_graph_flow.py:116]: Why not include the instances in the regression check?

File: tests/tests_integration/test_session/test_graph_flow.py [hunk 5]
```diff
@@ -76,12 +77,56 @@ def test_classic_to_dms(self, cognite_client: CogniteClient, data_regression: Da
...
+            }
+        )
+
+        assert len(instances) == 206
```
Comment (doctrino) [inline tests/tests_integration/test_session/test_graph_flow.py:120]: If you include the instances in the regression check, this line is not necessary.

File: tests/tests_unit/graph/test_loaders/test_dms_loader.py [hunk 6]
```diff
@@ -65,3 +65,4 @@ def test_imf_attribute_nodes():
 
     assert len(knowledge_nodes) == 56
     assert knowledge_nodes[0].sources[0].properties["predicate"].startswith("http")
+    assert len(store.multi_type_instances) == 63
```
Comment (doctrino) [inline tests/tests_unit/graph/test_loaders/test_dms_loader.py:68]: These types of tests are typically hard to debug when they go wrong. Ideally, smaller example, use data regression, or test for exact output.

File: cognite/neat/_store/_base.py [hunk 7]
```diff
@@ -292,6 +301,11 @@ def read(
         if not self.rules:
             warnings.warn("Rules not found in graph store!", stacklevel=2)
             return
+        if self.multi_type_instances:
```
Comment (doctrino) [inline cognite/neat/_store/_base.py:304]: I realize that we have 3 read methods, causing some duplicated warnings and checks. Could this be pulled out in its own method? Something like `_validate_ready_for_read`

File: cognite/neat/_session/engine/_load.py [hunk 8]
```diff
@@ -37,6 +37,7 @@ def load_neat_engine(client: CogniteClient | None, location: Literal["newest", "
 
     candidates: dict[Version, Callable[[], Path]] = {}
     if location == "cache" and cache_dir.exists():
+        print(cache_dir)
```
Comment (doctrino) [inline cognite/neat/_session/engine/_load.py:40]: Debug print

File: cognite/neat/_session/_prepare.py [hunk 9]
```diff
@@ -50,6 +60,47 @@ def __init__(self, state: SessionState, verbose: bool) -> None:
...
+        - attach values of generic attributes to nodes
+        - create associations between nodes
+        - remove unused generic attributes
+        - remove associations between nodes that do not exist in the extracted graph
+        - remove edges to nodes that do not exist in the extracted graph
+
+        and therefore safeguard CDF from a bad graph
+        """
+
+        DEXPI = get_default_prefixes_and_namespaces()["dexpi"]
```
Comment (doctrino) [inline cognite/neat/_session/_prepare.py:76]: Not a global constant
```suggestion
        dexpi = get_default_prefixes_and_namespaces()["dexpi"]
```

File: cognite/neat/_graph/transformers/_prune_graph.py [hunk 10]
```diff
@@ -182,10 +184,74 @@ def __init__(
...
 
     def transform(self, graph: Graph) -> None:
-        for object_type in self.node_prune_types:
-            nodes_without_neighbours = list(graph.query(self._query_template.format(rdf_type=object_type)))
+        for type_ in self.node_prune_types:
+            for (subject,) in list(graph.query(self._query_template.format(rdf_type=type_))):  # type: ignore
+                graph.remove((subject, None, None))
+
+
+class PruneTypes(BaseTransformer):
```
Comment (doctrino) [inline cognite/neat/_graph/transformers/_prune_graph.py:192]: Should be used in neat.drop.instances.

**Review by doctrino — APPROVED** [review]
My main comment is the test. Otherwise good work.

File: cognite/neat/_session/engine/_load.py (hunk 8 above)
Comment (nikokaoja) [inline cognite/neat/_session/engine/_load.py:40]:
```suggestion
```

**Pull request merged**